### PR TITLE
Gather cluster-level metrics in driver

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
With this commit we gather cluster-level metrics in the driver instead of the
mechanic. As these metrics are gathered via API calls there is no need to gather
them on the very same machine were an Elasticsearch node is running. Instead, it
defines clearer boundaries in between these two components.

Relates #697

### CODE REVIEW

File: esrally/mechanic/mechanic.py [hunk 1]
```diff
@@ -616,26 +490,20 @@ def receiveUnrecognizedMessage(self, msg, sender):
...
-                self.send(sender, BenchmarkStopped(self.metrics_store.to_externalizable(clear=True)))
             elif isinstance(msg, StopNodes):
                 self.logger.info("Stopping nodes %s.", self.mechanic.nodes)
                 self.mechanic.stop_engine()
-                self.send(sender, NodesStopped(self.metrics_store.to_externalizable()))
+                self.metrics_store.flush(refresh=False)
+                self.send(sender, NodesStopped(self.metrics_store.to_externalizable(clear=True)))
                 # clear all state as the mechanic might get reused later
                 self.metrics_store.close()
+                # TODO: Run the reporter (StatsCalculator) here to calculate summary stats and safe them to the
```
Comment (dliappis) [inline esrally/mechanic/mechanic.py]: s/safe/save

File: esrally/driver/driver.py [hunk 2]
```diff
@@ -257,17 +274,26 @@ def _requirements(self, host):
...
+        self.cluster_details = cluster_details
+
     def on_prepare_track(self, preparators, cfg, track):
         self.children = preparators
         msg = PrepareTrack(cfg, track)
         for child in self.children:
             self.send(child, msg)
 
     def after_track_prepared(self):
+        cluster_version = self.cluster_details["version"] if self.cluster_details else {}
```
[the remark this answers is not in the document]
Comment (danielmitterdorfer) [reply esrally/driver/driver.py]: As discussed offline, let's keep it defensive.

File: docs/migrate.rst [hunk 3]
```diff
@@ -4,6 +4,11 @@ Migration Guide
 Migrating to Rally 1.4.0
 ------------------------
 
+Node details are omitted from race metadata
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Before Rally 1.4.0, the file ``race.json`` has contained node details (such as the number of cluster nodes or details about the nodes' operating system version) if Rally has provisioned the cluster. With this release, this information is now omitted. This change also applies to the indices ``rally-races*`` in case you have setup an Elasticsearch metrics store. We recommend to use user tags in case such information is important, e.g. for visualising results.
```
[the remark this answers is not in the document]
Comment (danielmitterdorfer) [reply docs/migrate.rst]: Thanks! Addressed in https://github.com/elastic/rally/pull/779/commits/49523c852d744fec4f67942147d6b34f8da3a9fe

File: esrally/driver/driver.py [hunk 4]
```diff
@@ -307,8 +334,29 @@ def receiveMsg_PrepareTrack(self, msg, sender):
...
 
+def wait_for_rest_layer(es, max_attempts=20):
+    for attempt in range(max_attempts):
+        import elasticsearch
+        try:
+            es.info()
+            return True
+        except elasticsearch.ConnectionError as e:
+            if "SSL: UNKNOWN_PROTOCOL" in str(e):
+                raise exceptions.SystemSetupError("Could not connect to cluster via https. Is this a https endpoint?", e)
```
[the remark this answers is not in the document]
Comment (danielmitterdorfer) [reply esrally/driver/driver.py]: Addressed in https://github.com/elastic/rally/pull/779/commits/bdb66bfbea29a08a8557376f130151142d83b62b

File: esrally/mechanic/mechanic.py (hunk 1 above)
Comment (danielmitterdorfer) [reply esrally/mechanic/mechanic.py]: Addressed in https://github.com/elastic/rally/pull/779/commits/bdb66bfbea29a08a8557376f130151142d83b62b

**Pull request merged**